cds-snc/forms-api
# feat: token bucket rate limiting

# Summary | Résumé

Should not be merged until V1 is released to production (unless we decide to include it in V1).

Context: https://github.com/cds-snc/platform-forms-client/issues/3947

- Implements token bucket rate limiter using two different bucket capacities
  - Low rate limit is 500 requests per minute 
  - High rate limit is 1000 requests per minute
- Renames and moves `cache` (in lib/utils) to `redisClientAdapter` (in lib/integration/redis). The rate limiter code uses Redis to get a value that will be set by the web application. Since there is no cache concept involved here I decided to rename the two cache related functions. This way we can use it everywhere we need to access Redis and it also simplifies the mocks in our unit tests.

# Test instructions

- Modify the `config.ts` file to lower the token bucket capacity for both `lowRateLimiterConfiguration` and `highRateLimiterConfiguration`
- Send request to the API and observe both the response headers (there should be some new `X-RateLimit` properties) and response HTTP code (will be 429 when you have reached the limit)
- If you want to enable the high capacity bucket token just add a new entry in Redis with the following key/value pair: `rate-limit:${formId}` => "high"
